modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksStorageUtils.java

```diff
@@ -0,0 +1,284 @@
+
+        LONG_ARRAY_HANDLE.set(bytes, 0, updateCounter);
+
+        bytes[Long.BYTES] = (byte) (value == TOMBSTONE ? 0 : 1);
+
+        System.arraycopy(value, 0, bytes, Long.BYTES + Byte.BYTES, value.length);
+
+        return bytes;
+    }
+
+    /**
+     * Gets a list of longs from the byte array of longs.
+     *
+     * @param bytes Byte array of longs.
+     * @return List of longs.
+     */
+    @NotNull
+    static List<Long> getAsLongs(byte[] bytes) {
+        // Value must be divisible by a size of a long, because it's a list of longs
+        assert (bytes.length % Long.BYTES) == 0;
+
+        int listSize = bytes.length / Long.BYTES;
+
+        var longs = new ArrayList<Long>(listSize);
+
+        for (int i = 0; i < listSize; i++) {
+            long l = (long) LONG_ARRAY_HANDLE.get(bytes, i * Long.BYTES);
+
+            longs.add(l);
+        }
+
+        return longs;
+    }
+
+    /**
+     * Add a long value to an array of longs that is represented by an array of bytes.
+     *
+     * @param bytes Byte array that represents an array of longs.
+     * @param value New long value.
+     * @return Byte array with a new value.
+     */
+    static byte @NotNull [] addLongToLongsByteArray(byte @Nullable [] bytes, long value) {
+        // Index to insert new value at
+        int idx = bytes != null ? bytes.length : 0;
+
+        // Allocate a one long size bigger array
+        int size = idx + Long.BYTES;
+
+        var newBytes = new byte[size];
+
+        if (bytes != null)
+            // Copy the current value
+            System.arraycopy(bytes, 0, newBytes, 0, bytes.length);
+
+        LONG_ARRAY_HANDLE.set(newBytes, idx, value);
+
+        return newBytes;
+    }
+
+    /**
+     * Iterates over the given iterator passing key-value pairs to the given consumer and checks the iterator's status
+     * afterwards.
+     *
+     * @param iterator Iterator.
+     * @param consumer Consumer of key-value pairs.
+     * @throws RocksDBException If failed.
+     */
+    static void forEach(RocksIterator iterator, RocksBiConsumer consumer) throws RocksDBException {
+        for (; iterator.isValid(); iterator.next())
+            consumer.accept(iterator.key(), iterator.value());
+
+        checkIterator(iterator);
+    }
+
+    /**
+     * Iterates over the given iterator testing key-value pairs with the given predicate and checks the iterator's status
+     * afterwards.
+     *
+     * @param iterator Iterator.
+     * @param consumer Consumer of key-value pairs.
+     * @return {@code true} if a matching key-value pair has been found, {@code false} otherwise.
+     * @throws RocksDBException If failed.
+     */
+    static boolean find(RocksIterator iterator, RocksBiPredicate consumer) throws RocksDBException {
+        for (; iterator.isValid(); iterator.next()) {
+            boolean result = consumer.test(iterator.key(), iterator.value());
+
+            if (result) {
+                checkIterator(iterator);
```

Comment:
are you sure we need to check the iterator here? I can't imagine why it can be invalid here

Reply:
Not sure, but I think we can get the values from the invalid iterator